### DISCUSSION TITLE
How to access cms content model from within a module?

### DISCUSSION THREAD

**Discussion by LRathje** [OrchardCMS/OrchardCore#8684] [Q&A]
Hi
I managed to get a decoupled cms running fine and can access the content created in the admin from the pages using the IOrchardHelper and razor helpers.
But I have a MVC project running as a module. This also works fine for its own datamodel and for creating views after activating the module in features for the main project.
But how do I access the same content created in the admin directly from my viewpages in the MVC project module? Here the razor helpers for IOrchardHelper doesn't seem to work.
Best regards
Lars

**Comment by lampersky** [comment]
there are many ways you can access existing content items within the module,
you can access them via content manager or via yessql session directly

maybe share an example what is not working for you, it will be a way easier to help you

**Reply by LRathje** [reply]
Thanks @lampersky 

In my main orchardcms module (eg application.Web) I succesfully use this on a razorpage:
@inject OrchardCore.IOrchardHelper Orchard
var index = await Orchard.GetContentItemByHandleAsync($"alias:{Slug}");
.. and then I can access content on index like: @index.DisplayText

In my mvc module (eg application.Module referenced by application.Web) I try the same in a viewpage, but can't get access to the razor helpers. I do get access to eg HttpContext, but then I don't know how to get to the orchard content from there:
@inject OrchardCore.IOrchardHelper Orchard
var index = await Orchard.HttpContext?????

I also tried to install the nuget packages OrchardCore.ContentManagement.Abstractions and OrchardCore.DisplayManagement.Abstractions in the application.Module and add this to the viewpage:
@using OrchardCore.DisplayManagement
@using OrchardCore.ContentManagement

But I still don't know how to progress from here to access my OrchardCMS content. I would also be happy to just access the content from my mvc controller (might even be preferred) and then send it to the viewpage through a viewmodel. But how??

Best regards
Lars

**Reply by lampersky** [reply]
please make sure you have **_ViewImports.cshtml** file in Views directory inside your module, in this file you should add necessary usings, add tag helpers etc.

here is example file from one of my modules:

```
@inherits OrchardCore.DisplayManagement.Razor.RazorPage<TModel>
@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
@addTagHelper *, OrchardCore.DisplayManagement
@addTagHelper *, OrchardCore.ResourceManagement

@using OrchardCore.ContentManagement;
@using OrchardCore.DisplayManagement.Views;
@using OrchardCore.Media.ViewModels;
```

other than that, please make sure you have all dependencies added to your module csproj

**Reply by lampersky** [reply]
@LRathje there are many ways to get your content items within the controller:

- you can query content item by id using content manager:
`await _contentManager.GetAsync(contentItemId)`
- you can query exisiting query (sql or lucene):
```
            var query = await _queryManager.GetQueryAsync("<SomeQueryName>");
            var parameters = new Dictionary<string, object>() { { "from", from }, { "size", size } };
            var results = await _queryManager.ExecuteQueryAsync(query, parameters);
```
- you can query for a type using index (via yessql) or query Index directly:
```
            var results = await _session
                .Query<MyCustomType, MyCustomTypeIndex>(x => 
                    x.UserId == userId
                )
                .ListAsync();
```

**Reply by LRathje** [reply]
Thanks again @lampersky - I will give these hints a go and see what I can come up with. Will return with some feedback :).
Thanks Lars

**Reply by LRathje** [reply]
I got the "await _contentManager.GetAsync(contentItemId)" method working both directly in the viewpage and from my controller to viewmodel. That's a great start thanks :)
